### PR TITLE
prov/verbs: add support large (> 1 GB) transfers for ep_rdm

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
This is an implementation of RNDV-READ protocol and RMA ops for messages >1GB (RC QP limitation).

**tagged**: the receiver reads message per 1GB-segment and counts completions, then
sends completion ack-message to the sender.
**rma**: the originator posts operations per 1GB-segment and counts completions, then generates entry in OFI completion queue.

Current limitation is max_mr_size which can be registered with ibv_reg_mr.

refactoring of postponed queue: removed prefix tagged because of reusage in RMA

@a-ilango

### CODE REVIEW

**Comment by shefty** [comment]
@a-ilango - please merge once things look good.  I took a quick glance, and things looked okay overall, but I didn't delve into low-level details.

File: prov/verbs/src/ep_rdm/verbs_rdm.h [hunk 1]
```diff
@@ -120,8 +120,9 @@ struct fi_ibv_rdm_tagged_rndv_header {
 	struct fi_ibv_rdm_header base;
 	uint64_t src_addr;
 	void *id;
-	int len;
+	size_t total_len;
```
Comment (shefty) [inline prov/verbs/src/ep_rdm/verbs_rdm.h]: Variable sized fields really shouldn't be part of a wire protocol.  I.e. pointers and size_t, which differ based on platform. uint64_t is a better option.

**Comment by a-ilango** [comment]
The patches look good to me. @evgeny-leksikov Please let us know when this is ready for merge after addressing the comments.

**Comment by evgeny-leksikov** [comment]
@a-ilango , @shefty variable sized fields are removed, thanks!

**Comment by evgeny-leksikov** [comment]
Looks like the tests are failed because of merged PR https://github.com/ofiwg/libfabric/pull/2109, PR https://github.com/ofiwg/libfabric/pull/2110 should fix it.

**Comment by evgeny-leksikov** [comment]
@a-ilango @shefty could somebody merge this PR if there are no other objections? My next planned PR has a conflict without these changes. Thanks in advance!

File: prov/verbs/src/ep_rdm/verbs_tagged_ep_rdm_states.c [hunk 2]
```diff
@@ -283,10 +283,10 @@ fi_ibv_rdm_tagged_rndv_rts_send_ready(struct fi_ibv_rdm_tagged_request *request,
 	header->base.tag = request->minfo.tag;
 	header->base.service_tag = 0;
 	header->total_len = request->len;
-	header->src_addr = (uint64_t) (uintptr_t) request->src_addr;
+	header->src_addr = (uintptr_t)request->src_addr;
 
-	header->id = request;
-	request->rndv.id = request;
+	header->id = (uintptr_t)request;
+	request->rndv.id = (uintptr_t)request;
```
Comment (a-ilango) [inline prov/verbs/src/ep_rdm/verbs_tagged_ep_rdm_states.c]: Shouldn't this be (uint64_t) request ? Ditto for other casts above and below.

**Comment by a-ilango** [comment]
@evgeny-leksikov Can you please trigger the travis-CI again by updating the commit? I think you could just update the commit date and push to trigger travis-CI. This is to make sure the patch itself doesn't have any issues. The travis build has failed and no tests have been run against this patch.

File: prov/verbs/src/ep_rdm/verbs_tagged_ep_rdm_states.c (hunk 2 above)
Comment (evgeny-leksikov) [inline prov/verbs/src/ep_rdm/verbs_tagged_ep_rdm_states.c]: @a-ilango now, all pointers in scope of this function castes to uintptr_t then assignments to the destination variable which has also unsigned type with size not less than uintptr_t. So, I would prefer to change everything to this form in the future, probably in separated PR with refactoring. What's your opinion?

**Comment by evgeny-leksikov** [comment]
@a-ilango I don't know how, but it was triggered without update.

File: prov/verbs/src/ep_rdm/verbs_tagged_ep_rdm_states.c (hunk 2 above)
Comment (shefty) [inline prov/verbs/src/ep_rdm/verbs_tagged_ep_rdm_states.c]: > > +	header->id = (uintptr_t)request;
> > +	request->rndv.id = (uintptr_t)request;
> 
> Shouldn't this be (uint64_t) request ? Ditto for other casts above and
> below.

If request is a pointer, the cast needs to be uintptr_t, or there will be errors on 32-bit builds.

Comment (a-ilango) [inline prov/verbs/src/ep_rdm/verbs_tagged_ep_rdm_states.c]: Okay. :+1:

**Pull request merged**